include/Transforms/Functional.h

```diff
@@ -0,0 +1,364 @@
+//===- Functional.h ---------------------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef MLIR_REWRITE_FUNCTIONAL_H
+#define MLIR_REWRITE_FUNCTIONAL_H
+
+#include "mlir/IR/PatternMatch.h"
+#include "llvm/ADT/STLExtras.h"
+
+namespace mlir {
+namespace functional {
+
+namespace detail {
+/// A simple pattern rewriter that implements no special logic.
+class SimpleRewriter : public PatternRewriter {
+public:
+  SimpleRewriter(MLIRContext *context) : PatternRewriter(context) {}
+};
+} // end namespace detail
+
+/// Apply a pattern directly on an operation.
+template <typename OpT, typename PatternT, typename... Args>
+auto applyAt(OpT op, PatternT &&pattern, Args &&...args) {
+  detail::SimpleRewriter rewriter(op->getContext());
+  rewriter.setInsertionPoint(op);
+  return pattern(op, rewriter, std::forward<Args>(args)...);
+}
+
+/// Given a scope, apply a pattern until the first successful match and return
+/// the result.
+template <typename PatternT, typename... Args>
+auto applyOnce(Operation *scope, PatternT &&pattern, Args &&...args) {
```

Comment:
you could have the scopeop directly and simplify a bit

Reply:
These functions are meant to work in general, and not specifically with `ScopeOp` (although, yes, the naming is confusing...). I intend to keep the functionality separate as long as `ScopeOp` is tied to `linalg_transform` (it doesn't have to; I might make it more first-class).